cpp/src/arrow/ipc/reader.cc

```diff
@@ -860,27 +861,40 @@ Status ReadTensor(const Message& message, std::shared_ptr<Tensor>* out) {
 
 namespace {
 
-Status ReadSparseCOOIndex(const flatbuf::SparseTensor* sparse_tensor, int64_t ndim,
-                          int64_t non_zero_length, io::RandomAccessFile* file,
-                          std::shared_ptr<SparseIndex>* out) {
+Status ReadSparseCOOIndex(const flatbuf::SparseTensor* sparse_tensor,
+                          const std::vector<int64_t>& shape, int64_t non_zero_length,
+                          io::RandomAccessFile* file, std::shared_ptr<SparseIndex>* out) {
   auto* sparse_index = sparse_tensor->sparseIndex_as_SparseTensorIndexCOO();
+
+  std::shared_ptr<DataType> indices_type;
+  RETURN_NOT_OK(internal::GetSparseCOOIndexMetadata(sparse_index, &indices_type));
+
   auto* indices_buffer = sparse_index->indicesBuffer();
   std::shared_ptr<Buffer> indices_data;
   RETURN_NOT_OK(
       file->ReadAt(indices_buffer->offset(), indices_buffer->length(), &indices_data));
-  std::vector<int64_t> shape({non_zero_length, ndim});
-  const int64_t elsize = sizeof(int64_t);
-  std::vector<int64_t> strides({elsize, elsize * non_zero_length});
+  std::vector<int64_t> indices_shape(
+      {non_zero_length, static_cast<int64_t>(shape.size())});
```

Comment:
Perhaps not related with this PR, but is `non_zero_length` validated somewhere? What if it's not consistent with `indices_data`?

Reply:
`non_zero_length` links to `indices_shape[0]` in `SparseCOOIndex`.  On the other words, if `non_zero_length` is inconsistent with `indices_data`, it means `indices_shape` is inconsistent with `indices_data`.

If we need to do validation somewhere, I guess the appropriate place is Tensor's constructor.

And, I think we can remove `non_zero_length` attribute from the format, and we should do it.  I'll file the issue in JIRA.